File: apps/desktop/src/lib/project.ts
import { invoke as tauriInvoke } from '@tauri-apps/api/core';
import { CURRENT_PROTOCOL_VERSION, type InvokeCommand } from './health';

const STABLE_ID_PATTERN = /^[0-9a-f]{64}$/;
const OBJECT_ID_PATTERN = /^(?:[0-9a-f]{40}|[0-9a-f]{64})$/;
const MAX_REFERENCE_LENGTH = 1_024;
const MAX_PATH_DISPLAY_LENGTH = 32_768;

export interface OpenProjectRequestV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
}

export type GitHeadV1 =
  | { kind: 'born'; objectId: string; reference: string | null }
  | { kind: 'unborn'; reference: string };

export interface ProjectSummaryV1 {
  head: GitHeadV1;
  repositoryId: string;
  worktreeId: string;
  worktreeRootDisplay: string;
}

export type OpenProjectResultV1 =
  { status: 'cancelled' } | { project: ProjectSummaryV1; status: 'opened' };

export interface OpenProjectResponseV1 {
  protocolVersion: typeof CURRENT_PROTOCOL_VERSION;
  result: OpenProjectResultV1;
}

const invokeThroughTauri: InvokeCommand = (command, arguments_) =>
  tauriInvoke<unknown>(command, arguments_);

export async function openProject(
  invokeCommand: InvokeCommand = invokeThroughTauri,
): Promise<OpenProjectResponseV1> {
  const request: OpenProjectRequestV1 = {
    protocolVersion: CURRENT_PROTOCOL_VERSION,
  };
  const payload = await invokeCommand('open_project', { request });
  return parseOpenProjectResponseV1(payload);
}

export function parseOpenProjectResponseV1(payload: unknown): OpenProjectResponseV1 {
  if (!isRecord(payload) || !hasExactKeys(payload, ['protocolVersion', 'result'])) {
    throw new Error('Project response does not match the V1 schema.');
  }
  if (payload.protocolVersion !== CURRENT_PROTOCOL_VERSION) {
    throw new Error('Project response uses an unsupported protocol version.');
  }

  return {
    protocolVersion: payload.protocolVersion,
    result: parseResult(payload.result),
  };
}

function parseResult(value: unknown): OpenProjectResultV1 {
  if (!isRecord(value) || typeof value.status !== 'string') {
    throw new Error('Project response contains an invalid result.');
  }
  if (value.status === 'cancelled' && hasExactKeys(value, ['status'])) {
    return { status: 'cancelled' };
  }
  if (value.status === 'opened' && hasExactKeys(value, ['project', 'status'])) {
    return { project: parseProject(value.project), status: 'opened' };
  }
  throw new Error('Project response contains an invalid result.');
}

function parseProject(value: unknown): ProjectSummaryV1 {
  if (
    !isRecord(value) ||
    !hasExactKeys(value, ['head', 'repositoryId', 'worktreeId', 'worktreeRootDisplay']) ||
    typeof value.repositoryId !== 'string' ||
    !STABLE_ID_PATTERN.test(value.repositoryId) ||
    typeof value.worktreeId !== 'string' ||
    !STABLE_ID_PATTERN.test(value.worktreeId) ||
    typeof value.worktreeRootDisplay !== 'string' ||
    value.worktreeRootDisplay.length === 0 ||
    value.worktreeRootDisplay.length > MAX_PATH_DISPLAY_LENGTH
  ) {
    throw new Error('Project response contains an invalid project identity.');
  }

  return {
    head: parseHead(value.head),
    repositoryId: value.repositoryId,
    worktreeId: value.worktreeId,
    worktreeRootDisplay: value.worktreeRootDisplay,
  };
}

function parseHead(value: unknown): GitHeadV1 {
  if (!isRecord(value) || typeof value.kind !== 'string') {
    throw new Error('Project response contains an invalid HEAD state.');
  }
  if (
    value.kind === 'born' &&
    hasExactKeys(value, ['kind', 'objectId', 'reference']) &&
    typeof value.objectId === 'string' &&
    OBJECT_ID_PATTERN.test(value.objectId) &&
    (value.reference === null || isReference(value.reference))
  ) {
    return { kind: 'born', objectId: value.objectId, reference: value.reference };
  }
  if (
    value.kind === 'unborn' &&
    hasExactKeys(value, ['kind', 'reference']) &&
    isReference(value.reference)
  ) {
    return { kind: 'unborn', reference: value.reference };
  }
  throw new Error('Project response contains an invalid HEAD state.');
}

function isReference(value: unknown): value is string {
  return (
    typeof value === 'string' &&
    value.startsWith('refs/') &&
    value.length <= MAX_REFERENCE_LENGTH &&
    !Array.from(value).some((character) => {
      const codePoint = character.codePointAt(0);
      return codePoint !== undefined && (codePoint <= 31 || codePoint === 127);
    })
  );
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function hasExactKeys(value: Record<string, unknown>, expected: string[]): boolean {
  const keys = Object.keys(value).sort();
  const sortedExpected = [...expected].sort();
  return (
    keys.length === sortedExpected.length &&
    keys.every((key, index) => key === sortedExpected[index])
  );
}

File: crates/a3-protocol/src/project.rs
use crate::ProtocolVersion;
use serde::{Deserialize, Serialize};

/// Strict input payload for the V1 project-open command.
#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct OpenProjectRequestV1 {
    protocol_version: ProtocolVersion,
}

impl OpenProjectRequestV1 {
    /// Creates a request for a specific protocol version.
    #[must_use]
    pub const fn new(protocol_version: ProtocolVersion) -> Self {
        Self { protocol_version }
    }

    /// Creates a request for the protocol version emitted by this build.
    #[must_use]
    pub const fn current() -> Self {
        Self::new(ProtocolVersion::CURRENT)
    }

    /// Returns the requested protocol version.
    #[must_use]
    pub const fn protocol_version(self) -> ProtocolVersion {
        self.protocol_version
    }
}

/// Versioned response from one native project selection.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct OpenProjectResponseV1 {
    protocol_version: ProtocolVersion,
    result: OpenProjectResultV1,
}

impl OpenProjectResponseV1 {
    /// Creates a response for a picker dismissed by the user.
    #[must_use]
    pub const fn cancelled() -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: OpenProjectResultV1::Cancelled,
        }
    }

    /// Creates a response for a safely inspected local worktree.
    #[must_use]
    pub const fn opened(project: ProjectSummaryV1) -> Self {
        Self {
            protocol_version: ProtocolVersion::CURRENT,
            result: OpenProjectResultV1::Opened { project },
        }
    }

    /// Returns the protocol version carried by this response.
    #[must_use]
    pub const fn protocol_version(&self) -> ProtocolVersion {
        self.protocol_version
    }

    /// Returns the native-selection result.
    #[must_use]
    pub const fn result(&self) -> &OpenProjectResultV1 {
        &self.result
    }
}

/// Mutually exclusive result of the project-open command.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "status")]
pub enum OpenProjectResultV1 {
    /// The user dismissed the native directory picker.
    Cancelled,
    /// A selected Git worktree was safely identified.
    Opened {
        /// Bounded protocol projection of the project identity.
        project: ProjectSummaryV1,
    },
}

/// Safe project identity projected to the untrusted WebView.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ProjectSummaryV1 {
    repository_id: String,
    worktree_id: String,
    worktree_root_display: String,
    head: GitHeadV1,
}

impl ProjectSummaryV1 {
    /// Creates a WebView-safe project summary from boundary primitives.
    #[must_use]
    pub fn new(
        repository_id: String,
        worktree_id: String,
        worktree_root_display: String,
        head: GitHeadV1,
    ) -> Self {
        Self {
            repository_id,
            worktree_id,
            worktree_root_display,
            head,
        }
    }

    /// Returns the lowercase repository identity digest.
    #[must_use]
    pub fn repository_id(&self) -> &str {
        &self.repository_id
    }

    /// Returns the lowercase worktree identity digest.
    #[must_use]
    pub fn worktree_id(&self) -> &str {
        &self.worktree_id
    }

    /// Returns the non-authoritative worktree path intended only for display.
    #[must_use]
    pub fn worktree_root_display(&self) -> &str {
        &self.worktree_root_display
    }

    /// Returns the observed Git HEAD projection.
    #[must_use]
    pub const fn head(&self) -> &GitHeadV1 {
        &self.head
    }
}

/// WebView-safe Git HEAD state.
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields, rename_all = "camelCase", tag = "kind")]
pub enum GitHeadV1 {
    /// HEAD resolves to an object, optionally through a branch reference.
    Born {
        /// Lowercase SHA-1 or SHA-256 object identity.
        object_id: String,
        /// Full branch reference, absent for detached HEAD.
        reference: Option<String>,
    },
    /// HEAD points at a branch with no commit yet.
    Unborn {
        /// Full branch reference that will receive the first commit.
        reference: String,
    },
}

#[cfg(test)]
mod tests {
    use super::{GitHeadV1, OpenProjectRequestV1, OpenProjectResponseV1, ProjectSummaryV1};
    use crate::ProtocolVersion;
    use serde_json::json;

    #[test]
    fn opened_project_response_has_stable_json_shape() -> Result<(), serde_json::Error> {
        let response = OpenProjectResponseV1::opened(ProjectSummaryV1::new(
            "11".repeat(32),
            "22".repeat(32),
            "C:\\worktree".to_owned(),
            GitHeadV1::Unborn {
                reference: "refs/heads/main".to_owned(),
            },
        ));

        assert_eq!(
            serde_json::to_value(&response)?,
            json!({
                "protocolVersion": 1,
                "result": {
                    "status": "opened",
                    "project": {
                        "repositoryId": "11".repeat(32),
                        "worktreeId": "22".repeat(32),
                        "worktreeRootDisplay": "C:\\worktree",
                        "head": {
                            "kind": "unborn",
                            "reference": "refs/heads/main"
                        }
                    }
                }
            })
        );
        assert_eq!(response.protocol_version(), ProtocolVersion::V1);
        Ok(())
    }

    #[test]
    fn project_request_rejects_unknown_fields() {
        let result = serde_json::from_value::<OpenProjectRequestV1>(json!({
            "protocolVersion": 1,
            "selectedPath": "untrusted"
        }));

        assert!(result.is_err());
    }
}
